src/router_core/connections.c

```diff
@@ -2433,24 +2405,78 @@ static void qdr_link_inbound_detach_CT(qdr_core_t *core, qdr_action_t *action, b
         addr->ref_count--;
         qdr_check_addr_CT(core, addr);
     }
+}
+
+
+static void qdr_link_inbound_detach_CT(qdr_core_t *core, qdr_action_t *action, bool discard)
+{
+    qdr_connection_t *conn  = safe_deref_qdr_connection_t(action->args.connection.conn);
+    qdr_link_t       *link  = safe_deref_qdr_link_t(action->args.connection.link);
+    qdr_error_t      *error = action->args.connection.error;
+
+    if (discard || !conn || !link) {
+        qdr_error_free(error);
+        return;
+    }
+
+    if (link->state & QDR_LINK_STATE_DETACH_RECVD)
+        return;
+
+    qd_log(LOG_ROUTER_CORE, QD_LOG_DEBUG,
+           "[C%"PRIu64"][L%"PRIu64"] qdr_link_inbound_detach_CT()",
+           conn->identity, link->identity);
+
+    link->state |= QDR_LINK_STATE_DETACH_RECVD;
+
+    const bool first_detach = (link->state & QDR_LINK_STATE_DETACH_SENT) == 0;
+
+    if (link->core_endpoint) {
+        qdrc_endpoint_do_detach_CT(core, link->core_endpoint, error, first_detach);
+        return;
+    }
+
+    qdr_link_process_detach(core, link, error);
+
+    if (first_detach) {
+        // Send response detach
+        qdr_link_outbound_detach_CT(core, link, 0, QDR_CONDITION_NONE);
+    }
 
     if (error)
         qdr_error_free(error);
 }
 
 
-/* invoked on core thread to signal that the I/O thread has sent the detach
- */
-static void qdr_link_detach_sent_CT(qdr_core_t *core, qdr_action_t *action, bool discard)
+static void qdr_link_closed_CT(qdr_core_t *core, qdr_action_t *action, bool discard)
 {
-    qdr_link_t *link = safe_deref_qdr_link_t(action->args.connection.link);
+    qdr_link_t *link  = safe_deref_qdr_link_t(action->args.connection.link);
+    bool forced_close = action->args.connection.forced_close;
 
     if (discard || !link)
         return;
 
-    link->detach_send_done = true;
-    if (link->conn && link->detach_received)
-        qdr_link_cleanup_protected_CT(core, link->conn, link, "Link detached");
+    if (forced_close) {
+        // The link has been forced closed rather than cleanly detached.
+        if ((link->state & QDR_LINK_STATE_DETACH_RECVD) == 0) {
+            // detach-related cleanup was not done - do it now
+            if (link->core_endpoint) {
+                bool first_detach = (link->state & QDR_LINK_STATE_DETACH_SENT) == 0;
+                qdrc_endpoint_do_detach_CT(core, link->core_endpoint, 0, first_detach);
+                return;
+            }
+
+            qd_log(LOG_ROUTER_CORE, QD_LOG_DEBUG,
+                   "[C%"PRIu64"][L%"PRIu64"] qdr_link_closed_CT(forced=%s) handle %s detach",
+                   link->conn->identity, link->identity, forced_close ? "YES" : "NO",
+                       (link->state & QDR_LINK_STATE_DETACH_SENT) == 0 ? "first" : "second");
+
+            qdr_link_process_detach(core, link, 0);
+        }
+    }
+
+    qdr_link_cleanup_protected_CT(core, link->conn, link,
```

Comment:
Can the `qdr_link_cleanup_protected_CT()` function be removed and its contents copied into `qdr_link_closed_CT()`

Reply:
Good eye - not quite yet but soon.  My next patch will (hopefully) rework that "protected" stuff to remove the whole "Is the core finished? Is the I/O finished? Can it be freed?" flags handshake.  Hopefully that qdr_link_cleanup_protected_CT() goes away and we simply have a qdr_link_free_CT() call (like you recommend below).

Stay tuned.